### ISSUE TITLE
Failed to checkpoint when TCP sockets are in state CLOSE_WAIT

### ISSUE THREAD

**Issue by MadMub** [xemul/criu#262]
I'm very new to this low level of programming and working so closely with the OS so bear with me. I am attempting to use docker 1.13 rc4 to checkpoint and restore a containerized c++ app (nor am I a c++ developer by trade). when trying to checkpoint i get the following
```
alex@alex-ubuntu-16:~$ sudo docker checkpoint create testapp cp7
Error response from daemon: Cannot checkpoint container testapp: rpc error: code = 2 desc = exit status 1: "criu failed: type NOTIFY errno 0\nlog file: /var/lib/docker/containers/f083d943c5f7494a3675b0a0681de96e034465721a2d7f18e539cf035815a77f/checkpoints/cp7/criu.work/dump.log\n"
```
then i check that log
```
...
(00.018891) Dumping opened files (pid: 4502)
(00.018896) ----------------------------------------
(00.018931) Sent msg to daemon 12 0 0
pie: 1: __fetched msg: 12 0 0
pie: 1: __sent ack msg: 12 12 0
pie: 1: Daemon waits for command
(00.019385) Wait for ack 12 on daemon socket
(00.019397) Fetched ack: 12 12 0
(00.019416) 4502 fdinfo 0: pos:                0 flags:                0/0
(00.019425) Dumping pipe 13 with id 0x26 pipe_id 0x67ad
(00.019468) Dumping data from pipe 0x67ad fd 13
(00.019521) fdinfo: type: 0x2 flags: 0/0 pos:        0 fd: 0
(00.019545) 4502 fdinfo 1: pos:                0 flags:                1/0
(00.019551) Dumping pipe 14 with id 0x27 pipe_id 0x67ae
(00.019556) fdinfo: type: 0x2 flags: 01/0 pos:        0 fd: 1
(00.019566) 4502 fdinfo 2: pos:                0 flags:                1/0
(00.019570) Dumping pipe 15 with id 0x28 pipe_id 0x67af
(00.019575) fdinfo: type: 0x2 flags: 01/0 pos:        0 fd: 2
(00.019583) 4502 fdinfo 4: pos:                0 flags:                2/0
(00.019590)     Searching for socket 6b3a (family 2.6)
(00.019599) No filter for socket
(00.019623) Dumping inet socket at 4
(00.019628)     Dumping: ino   0x6b3a family    2 type    1 port    54322 state 10 src_addr 0.0.0.0
(00.019633)     Dumped: family 2 type 1 proto 6 port 54322 state 10 src_addr 0.0.0.0
(00.019638) fdinfo: type: 0x4 flags: 02/0 pos:        0 fd: 4
(00.019652) 4502 fdinfo 5: pos:                0 flags:             4002/0
(00.019658)     Searching for socket 67fa (family 2.6)
(00.019662) Error (criu/sk-inet.c:202): Name resolved on unconnected socket
(00.019671) ----------------------------------------
(00.019681) Error (criu/cr-dump.c:1313): Dump files (pid: 4502) failed with -1
...
```
It appears to have failed on some sockets. What's strange is I could swear the app was completely idle. It turns out that the app was using libcurl for network and a bunch of sockets were left in state CLOSE_WAIT.
```
alex@alex-ubuntu-16:~$ sudo nsenter -t 4502 -n netstat
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        1      0 123.45.6.7:32818        555.12.3.45:http        CLOSE_WAIT 
tcp        1      0 123.45.6.7:32822        555.12.3.45:http        CLOSE_WAIT 
tcp        1      0 123.45.6.7:32820        555.12.3.45:http        CLOSE_WAIT 
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node   Path
```
I substituted in fake IPs ^
Any information is of great help. I performed this through docker's experimental checkpoint/restore integration. Is there any other "easy" way of using CRIU, could I invoke it manually on the container if that would give more control. The goal is to migrate the container from one host to another.

**Comment by adrianreber** [comment]
Just yesterday I was looking into that. It works with the latest Fedora rawhide with a 4.10 rc kernel and yesterday's criu-dev branch:

```
# ./zdtm.py run -f h -t zdtm/static/socket-tcp-close-wait
Checking feature tcp_half_closed
=== Run 1/1 ================ zdtm/static/socket-tcp-close-wait

================== Run zdtm/static/socket-tcp-close-wait in h ==================
Start test
Running zdtm/static/socket-tcp-close-wait.hook(--post-start)
./socket-tcp-close-wait --pidfile=socket-tcp-close-wait.pid --outfile=socket-tcp-close-wait.out
Running zdtm/static/socket-tcp-close-wait.hook(--pre-dump)
State      Recv-Q Send-Q                                                                    Local Address:Port                                                                                   Peer Address:Port              
LISTEN     0      1                                                                                     *:8880                                                                                              *:*                  
CLOSE-WAIT 1      0                                                                             127.0.0.1:8880                                                                                      127.0.0.1:50022              
ESTAB      0      0                                                                             127.0.0.1:50024                                                                                     127.0.0.1:8880               
ESTAB      0      0                                                                             127.0.0.1:8880                                                                                      127.0.0.1:50024              
FIN-WAIT-2 0      0                                                                             127.0.0.1:50022                                                                                     127.0.0.1:8880               
Run criu dump
Running zdtm/static/socket-tcp-close-wait.hook(--pre-restore)
Run criu restore
Running zdtm/static/socket-tcp-close-wait.hook(--post-restore)
State      Recv-Q Send-Q                                                                    Local Address:Port                                                                                   Peer Address:Port              
LISTEN     0      1                                                                                     *:8880                                                                                              *:*                  
CLOSE-WAIT 1      0                                                                             127.0.0.1:8880                                                                                      127.0.0.1:50022              
ESTAB      0      0                                                                             127.0.0.1:50024                                                                                     127.0.0.1:8880               
ESTAB      0      0                                                                             127.0.0.1:8880                                                                                      127.0.0.1:50024              
FIN-WAIT-2 0      0                                                                             127.0.0.1:50022                                                                                     127.0.0.1:8880               
Check TCP images
dump/zdtm/static/socket-tcp-close-wait/32/1/tcp-stream-110ac0.img
dump/zdtm/static/socket-tcp-close-wait/32/1/tcp-stream-110ac1.img
dump/zdtm/static/socket-tcp-close-wait/32/1/tcp-stream-110abf.img
Send the 15 signal to  32
Wait for zdtm/static/socket-tcp-close-wait(32) to die for 0.100000
Running zdtm/static/socket-tcp-close-wait.hook(--clean)
Removing dump/zdtm/static/socket-tcp-close-wait/32
================= Test zdtm/static/socket-tcp-close-wait PASS ==================
```

So trying the same checkpointing on the latest kernel and criu should work.

**Comment by xemul** [comment]
Thank you, @adrianreber :)
Cc @avagin 
@MadMub , presumably your kernel doesn't have support for these states. IIRC you need 4.4 or higher (Andrey will correct me)

**Comment by MadMub** [comment]
I checked the kernel version
```
alex@alex-ubuntu-16:~$ uname -a
Linux alex-ubuntu-16 4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
is it because it is a virtual machine? do I need to install some kernel extensions? The version of criu being used is 2.6 i believe, should I try a newer one?

**Comment by xemul** [comment]
@MadMub , yes, the support for such TCP states has appeared recently and hasn't been released yet. You'll have to get criu-dev branch and compile it.

**Comment by MadMub** [comment]
@xemul I got a bit further in the log than before with a criu-dev built from source, it seemed to die at tcp repair. I ran criu check --all and got the following:
```
alex@alex-ubuntu-16:~/git/criu$ sudo criu check --all
[sudo] password for alex: 
Warn  (criu/autofs.c:79): Failed to find pipe_ino option (old kernel?)
Error (criu/cr-check.c:950): The TCP_REPAIR_WINDOW option isn't supported.
Error (criu/cr-check.c:934): TCP_REPAIR can't be enabled for half-closed sockets
Error (criu/cr-check.c:851): autofs not supported.
Error (criu/cr-check.c:967): Userfaultfd missing essential features
Looks good but some kernel features are missing
which, depending on your process tree, may cause
dump or restore failure.
```
Which makes it clear to me my kernel isn't up to date enough (I am on 4.4 though). Any idea what kernel version I do need (I will install it manually) would 4.6 do? also do I need to configure the kernel at all (at this point I'm starting to get out of my depths).

**Comment by tswift242** [comment]
@xemul Will the support for TCP states like this be released in criu 2.10 (next release)? I've gotten that same "Name resolved on unconnected socket" error a bunch of times. Also it'd be great to get a confirmation on the kernel version needed to support these states. We're using 4.4.x at the moment.

**Comment by mrjohndcarter** [comment]
A confirmation that the TCP state support will be in next release and explicit kernel version would be a great holiday gift :)

**Comment by MadMub** [comment]
I did a fresh install of everything from Debian, i installed kernel 4.8 through jessie-backports. It still seems to crash when trying to do a TCP_REPAIR of a half-closed socket. criu check --all also reports still that I do not have this feature.
```
TCP_REPAIR can't be enabled for half-closed sockets
```
Is this something I can enable? @adrianreber mentioned it works for kernel 4.10, do I just need to sit tight until that releases? Any info greatly appreciated thank you all for your time and hard work!

**Comment by avagin** [comment]
https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=319b0534b9

We need this patch to dump half-closed sockets. The first version of the upstream kernel with this patch will be 4.10-rc1.

**Comment by xemul** [comment]
@tswift242 , @mrjohndcarter , we'll do our best to have this work released in 2.10 :)

**Comment by xemul** [comment]
Should have been fixed in 2.10. Kudos @avagin !
